pkg/volume/util/util.go

```diff
@@ -614,6 +616,22 @@ func FsUserFrom(pod *v1.Pod) *int64 {
 	return fsUser
 }
 
+func SetFsUserIfUsernsEnabled(f *FileProjection, fsUser *int64) {
+	// We are adding this function out of extra cautious, to not do this if
+	// the feature flag is not active.
+	// However, this really seems being extra cautious. Given how this is
+	// called, this only changes the owner to the user running in the pod or
+	// to the user running in the container if all containers in the pod
+	// have the same RunAsUser (see FsUserFrom(), callers use as param the
+	// result of FsUserFrom()).
+	// Making the owner be yourself in these select cases shouldn't cause
+	// any issues. But for extra safety, we are adding
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
```

Comment:
Should this also check if `pod.spec.hostUsers` is set? Otherwise when this feature is turned on by default, it will change the behavior for all pods. In the original feature which added uid support to projected service account tokens, we intentionally kept the scope minimal to reduce chances of breaking existing workloads.

It would couple this ability to the user ns feature. We could also consider separating it out into its own independent feature, as I think there is value for uid volume ownership outside of user ns.

For that matter, is enabling fsuser a must have for phase 1? Or would converting gid only, which is today's behavior, sufficient? 

cc @gnufied who was exploring uid volume support in the past

Reply:
@msau42 yes, great point. I was planning to discuss in SIG-node using a new feature gate for this exactly for this reason. I thought it was okay for the first alpha, as probably it would take some time to coordinate with other SIGs (maybe another KEP?), etc.

I think is _almost_ a must. If you use configmaps or something and specify the mode (like 0600), then we need to update the owner (that is what we do here) so the pod with userns activated can read the file.

The group won't be enough unless fsGroup is used (which is definitely not always).

Do you think this is ok for alpha that is disabled by default and we discuss it in the coming weeks?

Reply:
probably want some issue or todo to track.

Reply:
> Do you think this is ok for alpha that is disabled by default and we discuss it in the coming weeks?

Ensuring we have a plausible path to enabling this seems like it should happen before merge.

Reply:
It is a long standing issue in Kubernetes that we do not support UID ownership for volumes: https://github.com/kubernetes/kubernetes/issues/57923

My understanding is that it impacts specific workloads that require 0600 permissions, but most others are fine with fsgroup, so can we ensure that at least fsgroup works with this?

Since this is alpha and disabled by default without a specific API field to control this, I am ok with keeping it and discussing what would be the best option moving forward.

Reply:
@msau42 thanks! Yes, fsGroup works fine with this. I'll try to add an e2e test tomorrow about it (12am here now :))

Reply:
> > Do you think this is ok for alpha that is disabled by default and we discuss it in the coming weeks?
> 
> Ensuring we have a plausible path to enabling this seems like it should happen before merge.

@liggitt There is. The code will look the same, just probably a new feature gate will be added for that. What do you think is missing?

Reply:
@liggitt ping?

Reply:
>> Ensuring we have a plausible path to enabling this seems like it should happen before merge.
>
> @liggitt There is. The code will look the same, just probably a new feature gate will be added for that. What do you think is missing?

Adding another feature gate just shifts enablement one more step down the road, it doesn't explain how existing clusters will enable the feature safely once the feature graduates. Feature gates are intended to cover the development life cycle of a feature - they are not intended to be long-term APIs. As such, they are expected to be deprecated and removed after a feature becomes GA or is dropped.

Maybe we're meaning different things by "enabled" ... is this expecting to make all pods use user namespaces eventually? If not, @msau42's comment is relevant and still needs to be addressed:

> Should this also check if pod.spec.hostUsers is set? Otherwise when this feature is turned on by default, it will change the behavior for all pods

Reply:
We have been trying hard to avoid special casing different volume types with "fsuser" support and sig-storage wants to pursue that problem as something that will work for every volume type. 

Is `fsuser` change necessary for this KEP? Shouldn't the user namespace mapping just work with `fsgroup` - which is de-factor standard so far in Kubernetes to allow non-root users to read/write files anyways. cc @mrunalp @derekwaynecarr

Reply:
> It is a long standing issue in Kubernetes that we do not support UID ownership for volumes: https://github.com/kubernetes/kubernetes/issues/57923
>
> My understanding is that it impacts specific workloads that require 0600 permissions, but most others are fine with fsgroup, so can we ensure that at least fsgroup works with this?

The UID/fsGroup issue @msau42 referenced is a good example of the type of problem we shouldn't make _worse_ with new API fields.

I can't tell if this part of the discussion settled on 1) adding testing to make sure this new field works with all the uses of existing API fields we allow, or if it settled on saying 2) we should accept this new field *not* working with some uses of existing API fields we allow because we already have a standing issue with fields not working together.

My perspective is that we should require tests proving the field combinations we allow in concert with this new field actually work, and if we know some combinations don't work, we should prevent those combinations in validation.

Reply:
@liggitt @gnufied 

>Maybe we're meaning different things by "enabled" ... is this expecting to make all pods use user namespaces eventually? If not, @msau42's comment is relevant and still needs to be addressed:

No, I don't mean all pod use user namespaces eventually, I meant have the pod.spec.hostUsers field set to false.

In any case, I can change this to not user FsUser and just use fsGroup. I think this should address all concerns in this thread.

The reason I didn't use FSGroup and went to FsUser is exactly the same linked by @gnufied (#57923). There is a comment in the code in userns_manager.go about it too, but github won't load to link it. If we use FSGroup, then the file can never have 0600 permissions (it will always have permissions for the group), so some libraries and tools (like ssh) that enforce those permissions will fail.

But sure, I can change this to use FSGroup for now. And I can talk with @gnufied and sig-storage to make a plan and go back to use FSUser in a future release (probably under a different feature gate, etc.)

Let me know if doing this change to only use FSGroup now and remove the parts about FsUser will solve the concern (I'd like to avoid spending several hours on this, to later find you don't like it either).

Thanks!

Reply:
I'll defer to the storage reviewers on whether the fsGroup change makes sense. I understood the issue with this code to be that it was making a decision based solely on the feature gate, when it should consider the pod spec `hostUsers` field value, right?

Reply:
I met with @mrunalp and @gnufied. @gnufied said if we remove the FsUser bits (was commit `pkg/volume: Add support for userns phase I volumes`, completely removed now) and we use FSGroup, it is fine from the sig-storage POV.

I've just updated the PR to do that.